File: nodejs/s3-to-sqs.js
/**
 * Script to re-queue messages from an S3 bucket back to the original AWS SQS queue.
 * This allows for processing messages that were previously moved to a DLQ and stored in S3.
 */

const { SQSClient, SendMessageBatchCommand } = require('@aws-sdk/client-sqs');
const { GetObjectCommand, S3Client } = require('@aws-sdk/client-s3');
const { fromIni } = require("@aws-sdk/credential-provider-ini");

// Generic AWS configuration
const region = "your-region"; // Example: "us-east-1"
const credentials = fromIni({ profile: 'your-profile' }); // Profile in your AWS credentials file

// SQS and S3 configuration
const sqsClient = new SQSClient({ region, credentials });
const s3Client = new S3Client({ region, credentials });

// Replace these with your original SQS queue URL and S3 bucket details
const queueURL = 'your-queue-url'; // Example: 'https://sqs.your-region.amazonaws.com/your-account-id/your-queue-name'
const bucketName = 'your-bucket-name'; // Example: 'your-dlq-messages-bucket'
const filePath = 'your-file-path'; // Example: 'dlq-messages.txt'

// Function to read messages from S3
const readMessagesFromS3 = async () => {
  try {
    const data = await s3Client.send(new GetObjectCommand({
      Bucket: bucketName,
      Key: filePath,
    }));
    const bodyContents = await streamToString(data.Body);
    return bodyContents.split('\n').filter(message => message.trim() !== '');
  } catch (error) {
    console.error('Error reading from S3:', error);
    throw error;
  }
};

// Helper function to convert stream to string
function streamToString(stream) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    stream.on('data', (chunk) => chunks.push(chunk));
    stream.on('error', reject);
    stream.on('end', () => resolve(Buffer.concat(chunks).toString('utf-8')));
  });
}

// Function to re-queue messages to SQS
const reQueueMessages = async (messages) => {
  // SQS batch command can only handle up to 10 messages at a time
  for (let i = 0; i < messages.length; i += 10) {
    const batch = messages.slice(i, i + 10).map((msg, index) => ({
      Id: String(index),
      MessageBody: msg,
    }));

    const command = new SendMessageBatchCommand({
      QueueUrl: queueURL,
      Entries: batch
    });

    try {
      await sqsClient.send(command);
      console.log(`Successfully re-queued ${batch.length} messages`);
    } catch (error) {
      console.error('Error re-queuing messages:', error);
    }
  }
};

// Main function to process the re-queuing
const processReQueueing = async () => {
  const messages = await readMessagesFromS3();
  if (messages.length > 0) {
    await reQueueMessages(messages);
  } else {
    console.log('No messages to re-queue');
  }
};

processReQueueing(); // Start the process

File: python/s3_to_sqs.py
import boto3
from botocore.exceptions import NoCredentialsError

# Specify the AWS region
aws_region = 'your-region'

# Initialize Boto3 clients for SQS and S3 with an explicit region
sqs_client = boto3.client('sqs', region_name=aws_region)
s3_client = boto3.client('s3', region_name=aws_region)

# Specify your S3 bucket details and the original SQS queue URL
bucket_name = 'your-bucket-name'
file_key = 'your-filenamet'  # The S3 object key (file name)
queue_url = 'url-of-your-queue'

def clear_s3_file_content(bucket, key):
    """Clears the content of the specified S3 file by overwriting it with an empty string."""
    try:
        s3_client.put_object(Bucket=bucket, Key=key, Body='')
        print("File content cleared successfully.")
    except Exception as e:
        print(f"An error occurred while clearing the file content: {str(e)}")

def read_messages_from_s3(bucket, key):
    """Read messages from an S3 object."""
    try:
        s3_object = s3_client.get_object(Bucket=bucket, Key=key)
        body = s3_object['Body'].read().decode('utf-8')
        messages = body.split('\n')
        return messages
    except NoCredentialsError:
        print("Error: AWS credentials not found.")
        return []
    except Exception as e:
        print(f"An error occurred: {str(e)}")
        return []

def requeue_messages(queue_url, messages):
    """Re-queue messages to the original SQS queue."""
    for message_body in messages:
        if not message_body.strip():
            continue  # Skip empty lines
        try:
            sqs_client.send_message(QueueUrl=queue_url, MessageBody=message_body)
            print("Message re-queued successfully.")
        except NoCredentialsError:
            print("Error: AWS credentials not found.")
        except Exception as e:
            print(f"An error occurred: {str(e)}")

def process_messages():
    """Main function to process messages from S3 to SQS."""
    messages = read_messages_from_s3(bucket_name, file_key)
    if messages:
        requeue_messages(queue_url, messages)
        clear_s3_file_content(bucket_name, file_key)
    else:
        print("No messages found in the specified S3 object.")

if __name__ == "__main__":
    process_messages()
